dask/array/routines.py

```diff
@@ -929,6 +932,304 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
+    ... ]
+    >>> h, edges = da.histogramdd(x, bins=edges)
+    >>> result = h.compute()
+    >>> result.shape
+    (4, 5, 3)
+
+    Defining the bins by total number and their ranges, along with
+    using weights:
+
+    >>> bins = (4, 5, 3)
+    >>> ranges = ((0, 1),) * 3  # expands to ((0, 1), (0, 1), (0, 1))
+    >>> w = da.random.uniform(0, 1, size=(1000,), chunks=x.chunksize[0])
+    >>> h, edges = da.histogramdd(x, bins=bins, range=ranges, weights=w)
+    >>> np.isclose(h.sum().compute(), w.sum().compute())
+    True
+
+    """
+
+    # logic used in numpy.histogramdd to handle normed/density.
+    if normed is None:
+        if density is None:
+            density = False
+    elif density is None:
+        # an explicit normed argument was passed, alias it to the new name
+        density = normed
+    else:
+        raise TypeError("Cannot specify both 'normed' and 'density'")
+
+    # generate token and name for task
+    token = tokenize(sample, bins, range, weights, density)
+    name = f"histogramdd-sum-{token}"
+
+    # N == total number of samples
+    # D == total number of dimensions
+    try:
+        # Recommended input ND-array
+        N, D = sample.shape
+    except (AttributeError, ValueError):
+        # If we have a sequence of 1D arrays
+        sample = atleast_2d(sample).T
+        N, D = sample.shape
+        # rechunk if necessary
+        if sample.chunksize[1] != D:
+            sample = sample.rechunk((sample.chunksize[1], D))
+
+    # Require only Array or Delayed objects for bins, range, and weights.
+    for argname, val in [("bins", bins), ("range", range), ("weights", weights)]:
+        if not isinstance(bins, (Array, Delayed)) and is_dask_collection(bins):
+            raise TypeError(
+                "Dask types besides Array and Delayed are not supported "
+                "for `histogramdd`. For argument `{}`, got: {!r}".format(argname, val)
+            )
+
+    # Require data to be chunked along the first axis only.
+    if sample.shape[1:] != sample.chunksize[1:]:
+        raise ValueError("Input array can only be chunked along the 0th axis")
+
+    # Require that the chunking of the sample and weights are compatible.
+    if weights is not None and weights.chunks[0] != sample.chunks[0]:
+        raise ValueError(
+            "Input array and weights must have the same shape "
+            "and chunk structure along the first dimension."
+        )
+
+    # if bins is a list, tuple, or array_like then make sure the
+    # length is the same as the number dimensions.
+    if isinstance(bins, (list, tuple)) or is_arraylike(bins):
+        if len(bins) != D:
+            raise ValueError(
+                "The dimension of bins must be equal to the dimension of the sample."
+            )
+    else:
+        # if not then bins must be a single scalar (possibly delayed),
+        # we expand it to the total number of dimensions.
+        bins = (bins,) * D
```

Comment:
Should we go ahead and raise here if it is a Dask collection given they are not supported later?